core/src/main/java/org/apache/brooklyn/enricher/stock/AbstractTransformer.java

```diff
@@ -59,27 +72,55 @@ public void setEntity(EntityLocal entity) {
+        if (targetSensor == null) {
+            throw new IllegalArgumentException("Enricher "+JavaClassNames.simpleClassName(this)+" has no "+TARGET_SENSOR.getName()+", and it cannot be inferred as "+SOURCE_SENSOR.getName()+" is also not set");
+        }
+        if (sourceSensor == null && triggerSensors.isEmpty()) {
+            throw new IllegalArgumentException("Enricher "+JavaClassNames.simpleClassName(this)+" has no "+SOURCE_SENSOR.getName()+" and no "+TRIGGER_SENSORS.getName());
+        }
+        if (producer.equals(entity) && (targetSensor.equals(sourceSensor) || triggerSensors.contains(targetSensor))) {
             // We cannot call getTransformation() here to log the tranformation, as it will attempt
             // to resolve the transformation, which will cause the entity initialization thread to block
             LOG.error("Refusing to add an enricher which reads and publishes on the same sensor: "+
-                producer+"."+sourceSensor+" (computing transformation with "+JavaClassNames.simpleClassName(this)+")");
+                producer+"->"+targetSensor+" (computing transformation with "+JavaClassNames.simpleClassName(this)+")");
             // we don't throw because this error may manifest itself after a lengthy deployment, 
             // and failing it at that point simply because of an enricher is not very pleasant
             // (at least not until we have good re-run support across the board)
             return;
         }
         
-        subscriptions().subscribe(producer, sourceSensor, this);
+        if (sourceSensor != null) {
+            subscriptions().subscribe(MutableMap.of("notifyOfInitialValue", true), producer, sourceSensor, this);
+        }
         
-        if (sourceSensor instanceof AttributeSensor) {
-            Object value = producer.getAttribute((AttributeSensor<?>)sourceSensor);
-            // TODO would be useful to have a convenience to "subscribeAndThenIfItIsAlreadySetRunItOnce"
-            if (value!=null) {
-                onEvent(new BasicSensorEvent(sourceSensor, producer, value, -1));
+        if (triggerSensors.size() > 0) {
+            SensorEventListener<Object> triggerListener = new SensorEventListener<Object>() {
+                @Override public void onEvent(SensorEvent<Object> event) {
+                    if (sourceSensor != null) {
+                        // Simulate an event, as though our sourceSensor changed
+                        Object value = producer.getAttribute((AttributeSensor<?>)sourceSensor);
+                        AbstractTransformer.this.onEvent(new BasicSensorEvent(sourceSensor, producer, value, event.getTimestamp()));
+                    } else {
+                        // Dangerous casting, but will assume that the transform doesn't care - otherwise 
+                        // it would have declared a sourceSensor!
+                        AbstractTransformer.this.onEvent((SensorEvent<T>)event);
```

Comment:
Personal preference - I'd consider passing null values here for sensor & value instead. If `sourceSensor` is null it would make more sense to fail with a NPE for the case where the transformer expects a specific sensor.